### PR TITLE
 refactor(refresh token): add refresh scheme

### PR DESCRIPTION
**Pull request by JoaoPedroAS51**

### CODE REVIEW

File: lib/schemes/refresh.js [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    this.isRefreshing = false
+    this.failedRequestQueue = []
+  }
+
+  // ---------------------------------------------------------------
+  // Token Expiration helpers
+  // ---------------------------------------------------------------
+
+  _getExpiration () {
+    const _key = this.$auth.options.token_expiration.prefix + this.name
```
Comment (pi0) [inline lib/schemes/refresh.js]: Why not using strategy/scheme options (`this.options`) for options? We can apply defaults in the constructor.

File: lib/schemes/refresh.js [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    const _key = this.$auth.options.refresh_token.prefix + this.name
+
+    return this.$auth.$storage.syncUniversal(_key)
+  }
+
+  async _updateTokens (result) {
+    const accessToken = getProp(result, this.options.endpoints.login.propertyName)
+
+    // Extract refresh token and set expiration
+    var tokenCreatedAt = getProp(result, this.options.endpoints.refresh.createdAt) || Date.now()
```
Comment (pi0) [inline lib/schemes/refresh.js:64]: Would you please use either `var` or `let`?

File: lib/schemes/refresh.js [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    var refreshToken = getProp(result, this.options.endpoints.refresh.token)
+    var defaultExpirationTime = parseInt(this.options.endpoints.refresh.defaultExpirationTime) || 1800
+    var tokenExpiration = getProp(result, this.options.endpoints.refresh.expiresIn)
+    var clientId = getProp(result, this.options.endpoints.refresh.clientId)
+
+    if (tokenExpiration) {
+      tokenExpiration = tokenCreatedAt + (tokenExpiration * 1000)
+    } else if (defaultExpirationTime && typeof defaultExpirationTime === 'number') {
```
Comment (pi0) [inline lib/schemes/refresh.js]: We already use `parseInt`. Is it required to check `typeof`?

File: lib/schemes/refresh.js [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    this.refreshInterval = setInterval(function () {
+      self._tokenRefresh()
+    }, intervalDuration)
+  }
+
+  _initializeErrorInterceptor () {
+    this.$auth.ctx.app.$axios.onError(error => {
+      // If error has config and response and error status is 401
+      if (error.config && error.response && error.response.status === 401) {
+        var token = this.$auth.getToken(this.name)
```
Comment (pi0) [inline lib/schemes/refresh.js]: Please use `const`

File: lib/schemes/refresh.js [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        if (token && headerToken && headerToken === token) {
+          // Add the failed request config to failedRequestQueue
+          if (!error.config.__isRetryRequest) {
+            error.config.__isRetryRequest = true
+            this.failedRequestQueue.push(error.config)
+          }
+
+          // If is not refreshing the token
+          if (!this.isRefreshing) {
+            var self = this
```
Comment (pi0) [inline lib/schemes/refresh.js]: Why `self` is used? Using an arrow function we can use `this`

File: lib/schemes/refresh.js [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    }
+
+    return this.$auth.fetchUserOnce().then((response) => {
+      // Initialize axios error interceptor
+      this._initializeErrorInterceptor()
+
+      // Only refresh token if user is logged in and is client side
+      if (process.client && this.$auth.loggedIn) {
+        var self = this
+        setTimeout(function () {
```
Comment (pi0) [inline lib/schemes/refresh.js]: Please prefer arrow function / self

File: lib/schemes/refresh.js (hunk 3 above)
Comment (JoaoPedroAS51) [reply lib/schemes/refresh.js]: https://github.com/JoaoPedroAS51/auth-module/pull/8/commits/b6b638a4cd016ceffa8ca4b55d656c6e29e25719

File: lib/schemes/refresh.js (hunk 2 above)
Comment (JoaoPedroAS51) [reply lib/schemes/refresh.js:64]: https://github.com/JoaoPedroAS51/auth-module/pull/8/commits/226e6d853dde76a2215deb16551a029ff2d17283

File: lib/schemes/refresh.js (hunk 4 above)
Comment (JoaoPedroAS51) [reply lib/schemes/refresh.js]: https://github.com/JoaoPedroAS51/auth-module/pull/8/commits/e49a94358eb76c04e490eacee5df71eef1a6459f

Comment (JoaoPedroAS51) [reply lib/schemes/refresh.js]: Sorry, I wrote wrong. var to const*

File: lib/schemes/refresh.js (hunk 5 above)
Comment (JoaoPedroAS51) [reply lib/schemes/refresh.js]: https://github.com/JoaoPedroAS51/auth-module/pull/8/commits/d91078ad98c7cb94cabceb211ba7591dd3688b5f

File: lib/schemes/refresh.js (hunk 6 above)
Comment (JoaoPedroAS51) [reply lib/schemes/refresh.js]: https://github.com/JoaoPedroAS51/auth-module/pull/8/commits/d91078ad98c7cb94cabceb211ba7591dd3688b5f

File: lib/schemes/refresh.js (hunk 1 above)
Comment (JoaoPedroAS51) [reply lib/schemes/refresh.js]: How do I apply the defaults?

Comment (pi0) [reply lib/schemes/refresh.js]: One solution:

````js
super(auth, {
   ...DEFAULTS,
   ...options
})
```
  
For deep assigning, we can use `defu` npm package. But I suggest flattening options like `refrashPrefix`

Comment (JoaoPedroAS51) [reply lib/schemes/refresh.js]: Thanks! https://github.com/JoaoPedroAS51/auth-module/pull/8/commits/24eba6c762621ecba04390c00461286f2054bdaf

**Pull request merged**